Adding 3-4-1

MichaelGNairn opened the pull request.

MichaelGNairn commented:
No history due to removal of sensitive file. Lizzie did automation script initially.

EmmaWoodONS commented on 3-4-1/README.md:
```diff
@@ -0,0 +1,78 @@
+
+### Code edits that may be needed: ###  
+*Note to QA: despite breaking up data link reigon data is still over 25000 lines so gets cut off, I think someone now has a NOMIS account to remove this issue however unsure so this would need fixing before fully complete. Also, when making this update  I followed how it was done on platform i.e starting region from 30 yrs and over however on reflection probably no need to do this now automated p ...
+  
```
The Regions issue has let to the biggest edit required in the code (see comments in example_config and compile_tables) - We would need 3 extra links to make it work - region totals, region female, and region male.

Even with a login it looks like the csv download option only goes up to 25000 rows. To get the full download you would need to use one of the other download types (I think). It may also cause issues if someone forgets to sign in to nomis before running the code. I think it is safer (and easier) to stick with the csv, and instead of splitting the links by sex, split them by year, and loop through the years in the compile_tables script (see comments on config and compile tables)

MichaelGNairn replied:
Yes, this makes sense. When working with NOMIS in Subnational it was a common issue that people just weren't signing into NOMIS (easily done) so much safer to ensure we "split" links to <25000. Looping by year would likely work for all indicators as well, so a catch-all, safe solution.

EmmaWoodONS commented on 3-4-1/example_config.R:
```diff
@@ -0,0 +1,21 @@
+
+#Link gives all totals excluding regional data 
+total_link <- "https://www.nomisweb.co.uk/api/v01/dataset/NM_161_1.data.csv?geography=2092957715,2092957699,2092957700,2092957703,2092957716&cause_of_death=9700,560,9,1130,1140,1150,1160,1170,1180,11095,11096,11098&gender=0&age=0...20&measure=2&measures=20100"
+
```
I think this link is missing covid (LC47) - is it meant to?
Because Regions doesn't fit in the 25000 row limit, we would need 3 new links (region total, then by male and female)

I think a better approach is to add e.g. `years < - c(2020, 2021)` as a configuration, then run the code once for each year (loop through the years list). Then just have two links e.g. `all_but_female_data` and `female_data` and don't specify date in the link (which is what happens when you select all years in the nomis selection anyway). "&date=", year is then added to the end of the link when looping through them (see comment in compile_tables)

MichaelGNairn replied:
Yes, makes sense and have done. 

Covid has recently been added to the UN metadata. Now included.

EmmaWoodONS commented on 3-4-1/compile_tables.R:
```diff
@@ -0,0 +1,44 @@
+
+library('stringr')
+library("dplyr")
+
```
put all library calls together (there are some in update_3-4-1.R)

MichaelGNairn replied:
Done

EmmaWoodONS commented on 3-4-1/compile_tables.R:
```diff
@@ -0,0 +1,44 @@
+
+existing_files <- list.files()
+output_folder_exists <- ifelse(output_folder %in% existing_files, TRUE, FALSE)
+
```
add a loop in here to go through the years. e.g.
```
# run one year of data at a time as the dataset is too big to download all years 
# at once
csv <- NULL

for (i in 1:length(years)) {
  all_but_female_link_temp <- paste0(all_but_female_link, "&date=", years[i])
  female_link_temp <- paste0(female_link, "&date=", years[i])
  
  source("update_3-4-1.R") 
  
  csv <- bind_rows(csv, csv_output)
}
```

MichaelGNairn replied:
Yes, this makes sense. Done.

EmmaWoodONS reviewed (dismissed):
The code doesn't quite work yet when running from main.  
I have made a few minor comments, but the main one is that Region causes issues with this way of splitting the data downloads. I suggest downloading each year separately and looping through the years in compile_tables (see comments in config and compile_tables).

Happy to chat it through if you disagree/ it is not clear! 

Once the region issue is resolved, I'll review again as this makes it a bit tricky reviewing the rest of the code properly.

MichaelGNairn commented on 3-4-1/compile_tables.R:
```diff
@@ -0,0 +1,44 @@
+source("example_config.R") # pulls in all the configurations
+source("compile_tables.R") # does the donkey-work of making the csv
+# at this point you should see lots of variables appear in the global environment 
+# pane (top right). These have been created by the update_type_1 script.
```
Done

MichaelGNairn commented on 3-4-1/compile_tables.R:
```diff
@@ -0,0 +1,44 @@
+# do not copy into your code as this is usually found in update_indicator_main.R
+
+source("example_config.R") # pulls in all the configurations
+source("compile_tables.R") # does the donkey-work of making the csv
```
DOne. This will be a relic of Lizzie and I running the indicator update line-by-line together.

MichaelGNairn commented on 3-4-1/update_3-4-1.R:
```diff
@@ -0,0 +1,145 @@
+library(tidyr)
+library(dplyr)
+library(stringr)
+library(tibble)
```
Done

MichaelGNairn commented on 3-4-1/example_config.R:
```diff
@@ -0,0 +1,21 @@
+# configurations for type 4 (Nomis) data
+
+# link for data - see the readme
+# Nomis link generation instructions should be explained in the README file
```
Done

MichaelGNairn commented on 3-4-1/README.md:
```diff
@@ -0,0 +1,78 @@
+   Rates: Age-standardised mortality rate only 
+   Sex: All 
+   Underlying Cause: Full ICD-10 Clasification > C00-C97, E10-E14, I00-I99,       J30-39, J40-J47, J60-J70, J80-J84, J85-J86, J90-94, J95, J96, J98 and LC47
+   
```
Done

MichaelGNairn commented on 3-4-1/README.md:
```diff
@@ -0,0 +1,78 @@
+The current nomis links were generated as follows:
+
+#### Mortality rate numbers ####
+From Nomis, select Query > Mortality statistics - underlying cause, sex and age.
```
Done.

MichaelGNairn commented:
Solved the region issue. Looping through years isn't working for me, hence data is only 2013 to 2016.

MichaelGNairn commented:
Tried a few things to try and get the looping years to work, but no success. Line 13-24 in compile_tables, with years being assigned on line 2 of config. @uzzelk any ideas? Maybe some fresh eyes may help? No problems if not!

EmmaWoodONS commented:
@MichaelGNairn I haven't tested it but I  found a couple of bugs, at least one of which is likely the issue:  
- In compile_tables get rid of the `source("update_3-4-1.R")` line that comes straight after `source("config.R")`  
- In update_3-4-1 change `read.csv(all_but_female_link)` to `read.csv(all_but_female_link_temp)` and the same with the female link  
My guess is that you should be able to do it with a single link without dropping females.  
  
Unrelated comment - put the library stuff in compile tables not in update_3-4-1:
```
library(SDGupdater)

packages <- c('tidyr', 'dplyr', 'stringr', 'tibble',
              'tools', 'ggplot2', 'DT', 'pander')

install_absent_packages(packages)
library(tidyr)
# etc
```

MichaelGNairn commented:
Think all is working now.  @uzzelk  would you like to check on your machine?

uzzelk commented:
> Think all is working now. @uzzelk would you like to check on your machine?

@MichaelGNairn I can check it Wednesday if that works? Currently running some confidence intervals code for quality jobs stuff!

uzzelk commented:
> Tried a few things to try and get the looping years to work, but no success. Line 13-24 in compile_tables, with years being assigned on line 2 of config. @uzzelk any ideas? Maybe some fresh eyes may help? No problems if not!
> 
> The NOMIS only bringing in 25000 rows could still be casuing the issue?

Sorry @MichaelGNairn I missed this!

MichaelGNairn commented:
> > Tried a few things to try and get the looping years to work, but no success. Line 13-24 in compile_tables, with years being assigned on line 2 of config. @uzzelk any ideas? Maybe some fresh eyes may help? No problems if not!
> > The NOMIS only bringing in 25000 rows could still be casuing the issue?
> 
> Sorry @MichaelGNairn I missed this!

No problems Katie. Loop was fine, I just called the "unlooped" dataframe in update_3-4-1 script.

MichaelGNairn commented:
> > Think all is working now. @uzzelk would you like to check on your machine?
> 
> @MichaelGNairn I can check it Wednesday if that works? Currently running some confidence intervals code for quality jobs stuff!

Yes, there's no rush. Hopefully have the QA.Rmd file working by then too.

MichaelGNairn commented:
@uzzelk forgot to say QA file for this is now done. I just hadn't pushed it from local to remote. 

Review at will, picking up what Emma started.

MichaelGNairn commented:
Hi @uzzelk , changes to ReadMe file made. 

Completely agree that the first sentence is confusing so I have edited it for clarity. I didn't actually understand what it meant, and I wrote the file! 

Sentence added to point 5 about making sure the years variable is correct when updating.

MichaelGNairn commented:
Compile tables issue addressed as well. This automation was started prior to the test_run feature, so it has slipped by. Thanks for spotting!

MichaelGNairn commented:
I have fixed the Wales being a region issue. NOMIS (and often the ONS) list Wales as a country and a region (hopefully see image). 
![image](https://user-images.githubusercontent.com/96240844/218050040-ff353258-3a56-4c9b-9814-e70d249ebf22.png)


I'm not even Welsh and this infuriates me!! In short, what has happened is all of the Wales data has been duplicated, as a region and a country. Country == Wales, Region == Blank. Country == Wales, Region == Wales. This should now be rectified.

Code in line 120 could be utilised by others. I have added to the Common Code Dictionary https://officenationalstatistics.sharepoint.com/:x:/r/sites/SusIneqSDG/Sample/Data_team/Projects_across_indicators/Automation/Common%20code%20dictionary.xlsx?d=wf6a995d735c84ae18da9090ee15966f0&csf=1&web=1&e=VVU29Y . 

It may struggle to run line by line because of the loop feature in compile tables. This loop is because there is so much data I had to bring in each year from NOMIS separately (NOMIS maxing out at 25000 rows).

MichaelGNairn commented:
No titles on a subplot means it is the headline value and therefore is a blank on the csv file (e.g. All ages). 
No data simply means there is no data. For example "Resident outside England and Wales" has only missing values by type of disease. 

Two data points per year issue is now resolved. I forgot to filter out different countries so it was plotting data for both England and Wales. 

In the age by chronic respiratory disease subtype graphs, the row with no disease sub-type has only old data. I think this is correct? It follows that when plotting disease subtypes there will be no data in the "no disease sub-type" plot? Seems to suggest to me that the old data should not be there!!

Have added in the rotating axis text code. have also added to common code dictionary. Very useful.

uzzelk commented:
Hi Michael,

Thanks for making the suggested changes. The README.md is much clearer to me now. The code runs and I can see Wales is no longer a region, however, there are now no values in for some of the disaggregations where there were two data point previously...  Specifically, in the 3-4-1_checks.knit output, there are no data points showing for:

- sex by type of disease
- age by chronic respiratory disease subtype
- sex by chronic respiratory disease subtype 

 I can see in the csv there are a lot of missing values but I'm not sure if they are true missing values or if something is going wrong with the code. As there were 2 data points for these graphs previously, I am guessing it's something that is happening in the code, but I've had a look and I can't work out what (sorry, I know that's not very helpful!). I'll let you take another look and hopefully it's something fairly easy to fix!

MichaelGNairn commented:
Changes made. The no data points in the three plots mentioned resulted from incorrect filtering when making the plots, meaning everything was filtered out. Hopefully this has now been resolved. Recommend a handful of spot checks for the csv with the source data directly on NOMIS.

uzzelk reviewed (approved):
QA output now appears to be working as expected. I have also spot checked a range of the values from the produced csv with values from the source data directly on NOMIS and they are all good. 

A final very, very minor comment re README - some of the points have full stops whereas others don't, and there are some random spaces (e.g., line 12 - space between update_indicator_main.R and the full stop and line 39, between codes I00-I99 and J30-39.

Other than that, this looks good to go!
